pkg/core/mempool/mempool.go

```diff
@@ -126,42 +154,80 @@ func NewMempool(eventBus *eventbus.EventBus, rpcBus *rpcbus.RPCBus, verifier tra
 	return m
 }
 
-// Run spawns the mempool lifecycle routine. The whole mempool cycle is around
-// getting input from the outside world (from input channels) and provide the
-// actual list of the verified txs (onto output channel).
-//
-// All operations are always executed in a single go-routine so no
-// protection-by-mutex needed.
+// Run spawns the mempool lifecycle routines.
 func (m *Mempool) Run(ctx context.Context) {
-	go func() {
-		ticker := time.NewTicker(idleTime)
-		defer ticker.Stop()
-
-		for {
-			select {
-			// rpcbus methods.
-			case r := <-m.sendTxChan:
-				go handleRequest(r, m.processSendMempoolTxRequest, "SendTx")
-			case r := <-m.getMempoolTxsChan:
-				handleRequest(r, m.processGetMempoolTxsRequest, "GetMempoolTxs")
-			case r := <-m.getMempoolTxsBySizeChan:
-				handleRequest(r, m.processGetMempoolTxsBySizeRequest, "GetMempoolTxsBySize")
-			case b := <-m.acceptedBlockChan:
-				m.onBlock(b)
-			case <-ticker.C:
-				m.onIdle()
-			// Mempool terminating.
-			case <-ctx.Done():
-				// m.eventBus.Unsubscribe(topics.Tx, m.txSubscriberID)
-				return
+	// Main Loop
+	go m.Loop(ctx)
+
+	// Loop to drain pendingPropagation and try to propagate transaction
+	go m.propagateLoop(ctx)
+}
+
+// Loop listens for GetMempoolTxs request and topics.AcceptedBlock events.
+func (m *Mempool) Loop(ctx context.Context) {
+	ticker := time.NewTicker(idleTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		// rpcbus methods.
+		case r := <-m.sendTxChan:
+			// TODO: This handler should be deleted once new wallet is integrated
+			go handleRequest(r, m.processSendMempoolTxRequest, "SendTx")
+		case r := <-m.getMempoolTxsChan:
+			handleRequest(r, m.processGetMempoolTxsRequest, "GetMempoolTxs")
+		case r := <-m.getMempoolTxsBySizeChan:
+			handleRequest(r, m.processGetMempoolTxsBySizeRequest, "GetMempoolTxsBySize")
+		case b := <-m.acceptedBlockChan:
+			m.onBlock(b)
+		case <-ticker.C:
+			m.onIdle()
+		case <-ctx.Done():
+			// Mempool terminating
+			return
+		}
+
+		ticker.Reset(idleTime)
+	}
+}
+
+func (m *Mempool) propagateLoop(ctx context.Context) {
+	for {
+		select {
+		case t := <-m.pendingPropagation:
+			// Ensure we propagate at proper rate
+			if m.limiter != nil {
+				if err := m.limiter.Wait(ctx); err != nil {
+					log.WithError(err).Error("failed to limit rate")
+				}
+			}
+
+			txid, err := t.tx.CalculateHash()
+			if err != nil {
+				log.WithError(err).Error("failed to calc hash")
+				continue
 			}
 
-			ticker.Reset(idleTime)
+			if config.Get().Kadcast.Enabled {
+				// Broadcast ful transaction data in kadcast
+				err = m.kadcastTx(t)
+			} else {
+				// Advertise the transaction hash to gossip network via "Inventory Vectors"
+				err = m.advertiseTx(txid)
+			}
+
+			if err != nil {
+				log.WithField("txid", hex.EncodeToString(txid)).WithError(err).Error("failed to propagate")
+			}
+
+		// Mempool terminating
+		case <-ctx.Done():
+			return
```

Comment:
Shouldn't we drain `m.pendingPropagation` in case of context cancelation?

Reply:
Not for now probably. Later on when we implement #1059 , we can drain and persist these txs as well.

Reply:
FYI
https://github.com/dusk-network/dusk-blockchain/issues/1059#issuecomment-1012049540